takari-builder/src/main/java/io/takari/builder/internal/maven/MavenDependencyResolver.java

```diff
@@ -0,0 +1,308 @@
+        result.remove(Artifact.SCOPE_RUNTIME);
+      } else if (Artifact.SCOPE_TEST.equals(scope)) {
+        result.clear();
+      }
+    }
+
+    return result;
+  }
+
+  public static Collection<String> toScopes(ResolutionScope scope) {
+    if (ResolutionScope.COMPILE.equals(scope)) {
+      return Arrays.asList(Artifact.SCOPE_COMPILE, Artifact.SCOPE_SYSTEM, Artifact.SCOPE_PROVIDED);
+    } else if (ResolutionScope.RUNTIME.equals(scope)) {
+      return Arrays.asList(Artifact.SCOPE_COMPILE, Artifact.SCOPE_RUNTIME);
+    } else if (ResolutionScope.COMPILE_PLUS_RUNTIME.equals(scope)) {
+      return Arrays.asList(Artifact.SCOPE_COMPILE, Artifact.SCOPE_SYSTEM, Artifact.SCOPE_PROVIDED,
+          Artifact.SCOPE_RUNTIME);
+    } else if (ResolutionScope.TEST.equals(scope)) {
+      return Arrays.asList(Artifact.SCOPE_COMPILE, Artifact.SCOPE_SYSTEM, Artifact.SCOPE_PROVIDED,
+          Artifact.SCOPE_RUNTIME, Artifact.SCOPE_TEST);
+    }
+    return Collections.emptyList();
+  }
+
+  private DependencyResolutionResult resolveDependencies(ResolutionScope scope,
+      DependencyFilter resolutionFilter) throws DependencyResolutionException {
+    ClassLoader tccl = Thread.currentThread().getContextClassLoader();
+    try {
+      ClassLoader projectRealm = project.getClassRealm();
+      if (projectRealm != null && projectRealm != tccl) {
+        Thread.currentThread().setContextClassLoader(projectRealm);
+      }
+
+      RequestTrace trace = RequestTrace.newChild(null, resolutionFilter);
+      ScopeBasedDependencyResolutionResult result = new ScopeBasedDependencyResolutionResult();
+      ArtifactTypeRegistry stereotypes = repoSession.getArtifactTypeRegistry();
+      CollectRequest collect = new CollectRequest();
+      collect.setRootArtifact(RepositoryUtils.toArtifact(project.getArtifact()));
+      collect.setRequestContext("projectScope");
+      collect.setRepositories(project.getRemoteProjectRepositories());
+
+      for (org.apache.maven.model.Dependency dependency : project.getDependencies()) {
+        if (StringUtils.isEmpty(dependency.getGroupId())
+            || StringUtils.isEmpty(dependency.getArtifactId())
+            || StringUtils.isEmpty(dependency.getVersion())) {
+          // guard against case where best-effort resolution for invalid models is requested
+          continue;
+        }
+        collect.addDependency(RepositoryUtils.toDependency(dependency, stereotypes));
+      }
+
+      DependencyManagement depMngt = project.getDependencyManagement();
+      if (depMngt != null) {
+        for (org.apache.maven.model.Dependency dependency : depMngt.getDependencies()) {
+          collect.addManagedDependency(RepositoryUtils.toDependency(dependency, stereotypes));
+        }
+      }
+
+      DependencyRequest depRequest = new DependencyRequest(collect, resolutionFilter);
+      depRequest.setTrace(trace);
+
+      collect.setTrace(RequestTrace.newChild(trace, depRequest));
```

Comment:
I think this collection/trace api can be cleaned up.... we have a pretty simple hierarchical timer system in core that could probably meet your needs and expose fewer classes.